frex/utils/constraint_solver_utils.py

```diff
@@ -0,0 +1,123 @@
+        containing (field_to_constrain, type_of_constraint (i.e., 'eq', 'leq', 'geq), value_to_constrain).
+        :return: A tuple of tuples, containing candidates assigned to each section.
+        """
+        solver = pywraplp.Solver.CreateSolver("SCIP")
+
+        candidate_count = len(candidates)
+        solve_choice = {}
+        for i in range(candidate_count):
+            for j in range(num_sections):
+                solve_choice[i, j] = solver.IntVar(0, 1, "")
+
+        # each item is only assigned to one section
+        for i in range(candidate_count):
+            solver.Add(
+                solver.Sum([solve_choice[i, j] for j in range(num_sections)]) <= 1
+            )
+
+        # each section has exactly per_section_count items chosen
+        for j in range(num_sections):
+            solver.Add(
+                solver.Sum([solve_choice[i, j] for i in range(candidate_count)])
+                == per_section_count
+            )
+
+        # constraints to apply on each section, based on certain field names
+        for psc in per_section_constraints:
+            for j in range(num_sections):
+                ss = solver.Sum(
+                    [
+                        getattr(candidates[i].domain_object, psc[0])
+                        * solve_choice[i, j]
+                        for i in range(candidate_count)
+                    ]
+                )
+                if psc[1] == "eq":
+                    solver.Add(ss == psc[2])
+                elif psc[1] == "leq":
+                    solver.Add(ss <= psc[2])
+                else:
+                    solver.Add(ss >= psc[2])
+
+        # constraints to apply to the overall solution, based on certain field names
+        for oc in overall_constraints:
+            ss = solver.Sum(
+                [
+                    getattr(candidates[i].domain_object, oc[0]) * solve_choice[i, j]
+                    for i in range(candidate_count)
+                    for j in range(num_sections)
+                ]
+            )
+            if oc[1] == "eq":
+                solver.Add(ss == oc[2])
+            elif oc[1] == "leq":
+                solver.Add(ss <= oc[2])
+            else:
+                solver.Add(ss >= oc[2])
+
+        # maximize score
+        objective_terms = []
+        for i in range(candidate_count):
+            for j in range(num_sections):
+                # in the future, maybe incorporate more into this objective function
+                # e.g., we would prefer the combination of all items in a section to have some field value in a range,
+                # so incorporate that into the score somehow?
+                objective_terms.append(candidates[i].total_score * solve_choice[i, j])
+        solver.Maximize(solver.Sum(objective_terms))
+
+        status = solver.Solve()
+
+        if status == pywraplp.Solver.OPTIMAL or status == pywraplp.Solver.FEASIBLE:
```

Comment:
I'd also like to see this return NamedTuple/@dataclass with the information you're throwing away, like the scores. It's easier to read than tuple[tuple[...]]. You may be able to use this info in explanations, too?